runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.nonrepudiation.api.v1
+
+import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server.Route
+import com.daml.nonrepudiation.{CommandIdString, SignedPayload, SignedPayloadRepository}
+import com.google.common.io.BaseEncoding
+import org.slf4j.{Logger, LoggerFactory}
+import spray.json.{DefaultJsonProtocol, RootJsonFormat}
+
+import scala.collection.compat.immutable.ArraySeq
+
+private[api] final class SignedPayloadsEndpoint private (
+    signedPayloads: SignedPayloadRepository.Read[CommandIdString]
+) extends SignedPayloadsEndpoint.JsonProtocol {
+
+  import SignedPayloadsEndpoint._
+
+  private val route: Route =
+    path(Segment.map(CommandIdString.wrap)) { commandId =>
+      get {
+        handleExceptions(logAndReport(logger)(UnableToRetrieveTheSignedPayload)) {
+          val responses = signedPayloads.get(commandId).map(toResponse)
+          if (responses.nonEmpty) {
+            complete(responses)
+          } else {
+            reject
+          }
+        }
+      }
+    }
+
+}
+
+object SignedPayloadsEndpoint {
+
+  def apply(signedPayloads: SignedPayloadRepository.Read[CommandIdString]): Route =
+    new SignedPayloadsEndpoint(signedPayloads).route
+
+  private val logger: Logger = LoggerFactory.getLogger(classOf[SignedPayloadsEndpoint])
+
+  private[api] val UnableToRetrieveTheSignedPayload: String =
+    "An error occurred when trying to retrieve the signed payload, please try again."
+
+  final case class Response(
+      algorithm: String,
+      fingerprint: String,
+      payload: String,
+      signature: String,
+      timestamp: Long,
+  )
+
+  private def base64Url(bytes: ArraySeq.ofByte): String =
+    BaseEncoding.base64Url().encode(bytes.unsafeArray)
+
+  def toResponse(signedPayload: SignedPayload): Response =
+    Response(
+      algorithm = signedPayload.algorithm,
+      fingerprint = base64Url(signedPayload.fingerprint),
+      payload = base64Url(signedPayload.payload),
+      signature = base64Url(signedPayload.signature),
+      timestamp = signedPayload.timestamp.toEpochMilli,
+    )
+
+  private[api] trait JsonProtocol extends SprayJsonSupport with DefaultJsonProtocol {
+
+    protected implicit val apiSignedPayloadFormat: RootJsonFormat[Response] =
```

Comment:
for the JSON API and the trigger service we follow a common format of `result` and `error` fields and a `status` field in all responses. Have you considered adapting that here as well? If not, what do you think of doing so? totally fine not to do this at all or do it in a separate PR.

Reply:
Sure, I was not aware of this. What is the `status` field though?

Reply:
It’s the status code. It makes more sense in the JSON API since annoyingly it can be different from the http status code in some cases (we stream some responses and we don’t know the status code when sending the header, fun side note: that’s why grpc uses http trailers instead of headers afaik).

Reply:
Ok.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/8982/commits/9faaed6d13deedf47fd8f07b7ddd4dfce425dcb7

I'm not 100% sure that we should include the status code in this API, but we can review the decision at a later point.

Following the `{ "error": "..." }`/`{ "result": { ... } }` pattern actually makes sense, it will hopefully allow users to reuse some code, or at least to be familiar with the output format.